File: api/matlab/Result.m
classdef Result < Entity
    %RESULT CMAC operation result.
    %   If not successful, additional message is provided of the failure.

    % Method types as uint32
    properties(Hidden, Constant)
        METHOD_DELETE uint32 = 0;
        METHOD_IS_SUCCESSFUL uint32 = 100;
        METHOD_GET_MESSAGE uint32 = 101;
    end

    methods
        function obj = Result(handle, mexname, deleteId)
            %RESULT Construct an instance of this class
            arguments
                handle uint64
                mexname string = "ResultMex";
                deleteId uint32 = Result.METHOD_DELETE
            end
            obj@Entity(mexname, handle, deleteId);
        end
        
        function success = IsSuccessful(obj)
            %ISSUCCESSFUL Indicates if the operation was successful.
            arguments
                obj Result
            end
            success = feval(obj.mexName ...
                , Result.METHOD_IS_SUCCESSFUL, obj.handle);
        end

        function msg = GetMessage(obj)
            %GETMESSAGE Acquires the message for any unsuccessful
            %operation.
            arguments
                obj Result
            end
            msg = feval(obj.mexName ...
                , Result.METHOD_GET_MESSAGE, obj.handle);
        end
    end
end

File: core/src/Result.cpp
#include "cmaclib/Result.h"

#include <utility>

using namespace CmacLib;

void Result::SetIsSuccessful(bool success)
{
	this->successful = success;
}

void Result::SetMessage(std::string&& msg)
{
	this->message = ::std::move(msg);
}

Result::Result(): successful(false), message(""){}

Result::~Result(){}

bool Result::IsSuccessful()
{
	return this->successful;
}

std::string Result::GetMessage()
{
	return this->message;
}

/// @brief Indicates if an operation was successful.
/// @param pResult The pointer to the result instance.
/// @return Success.
cmaclib_bool_t CCmacLibIResultIsSuccessful(CCmacLibIResult* pResult) {
	return pResult->IsSuccessful();
}

/// @brief If not successful, a failure message is provided.
/// @param pResult The pointer to the result instance.
const char* CCmacLibIResultGetMessage(CCmacLibIResult* pResult) {
	return pResult->GetMessage().c_str();
}

/// @brief Destructor.
/// @param pResult The pointer to the result instance.
void CCmacLibIResultDestory(CCmacLibIResult* pResult) {
	if (pResult) delete pResult;
}